cpp/include/cudf/strings/string_view.cuh

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstddef>
+#include <cuda_runtime.h>
+#include <cudf/types.hpp>
+
+namespace cudf
+{
+
+// utf8 characters are 1-4 bytes
+typedef unsigned int Char;
+
+/**---------------------------------------------------------------------------*
+ * @brief A non-owning, immutable view of device data that is variable length
+ * character array representing a UTF-8 string. The caller must maintain the
+ * device memory for the lifetime of this instance.
+ *
+ * It provides a simple wrapper and string operations for individual char array
+ * within a strings column. This is likely created dynamically and temporarily.
+ * It is not recommended to be allocated directly on the global memory heap.
+ *---------------------------------------------------------------------------**/
+class string_view
+{
+ public:
+  string_view() = default;
+  /**---------------------------------------------------------------------------*
+   * @brief Create instance from existing device char array.
+   *
+   * @param data Device char array encoded in UTF8.
+   * @param bytes Number of bytes in data array.
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ string_view(const char* data, size_type bytes);
+  /**---------------------------------------------------------------------------*
+   * @brief Create instance from existing device char array. The array must
+   * include a null-terminator ('\0).
+   *
+   * @param data Device char array encoded in UTF8.
+   *---------------------------------------------------------------------------**/
+  __device__ string_view(const char* data);
+  string_view(const string_view&) = default;
+  string_view(string_view&&) = default;
+  ~string_view() = default;
+  string_view& operator=(const string_view&) = default;
+  string_view& operator=(string_view&&) = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Return the number of bytes in this string
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ size_type size() const;
+  /**---------------------------------------------------------------------------*
+   * @brief Return the number of bytes in this string
```

Comment:
The `size()` and `length()` are the same number because the C++ string classes do not have variable length characters. So I'd like to deviate here and make `size()` the number of bytes and `length()` to be the number of characters. The aligns `size()` value for the chars column child in a strings column as the total number of bytes.

Reply:
Could we do `size_bytes()` to avoid confusion?